frontend/src/metabase/visualizations/echarts/cartesian/model/series.ts

```diff
@@ -324,6 +330,225 @@ function shouldRenderCompact(
 }
 
+export function getWaterfallChartDataDensity(
+  dataset: ChartDataset,
+  waterfallLabelFormatter: RawValueFormatter | undefined,
+  settings: ComputedVisualizationSettings,
+  renderingContext: RenderingContext,
+): WaterFallChartDataDensity {
+  const type = "waterfall";
+  if (
+    !settings["graph.show_values"] ||
+    settings["graph.label_value_frequency"] === "all"
+  ) {
+    return {
+      type,
+      averageLabelWidth: 0,
+      totalNumberOfLabels: 0,
+    };
+  }
+
+  let totalNumberOfLabels = 0;
+  let sumOfLabelWidths = 0;
+
+  const fontStyle = {
+    family: renderingContext.fontFamily,
+    weight: CHART_STYLE.seriesLabels.weight,
+    size: CHART_STYLE.seriesLabels.size,
+  };
+
+  dataset.forEach(datum => {
+    const value = datum[WATERFALL_VALUE_KEY];
+
+    if (value == null) {
+      return;
+    }
+
+    totalNumberOfLabels += 1;
+
+    if (!waterfallLabelFormatter) {
+      return;
+    }
+
+    const labelWidth = renderingContext.measureText(
+      waterfallLabelFormatter(value),
+      fontStyle,
+    );
+
+    sumOfLabelWidths += labelWidth;
+  });
+
+  const averageLabelWidth =
+    totalNumberOfLabels > 0 ? sumOfLabelWidths / totalNumberOfLabels : 0;
+
+  return {
+    type,
+    averageLabelWidth,
+    totalNumberOfLabels,
+  };
+}
+
+export function getCartesianChartDataDensity(
+  seriesModels: SeriesModel[],
+  stackModels: StackModel[],
+  dataset: ChartDataset,
+  seriesLabelsFormatters: SeriesFormatters,
+  stackedLabelsFormatters: StackedSeriesFormatters,
+  settings: ComputedVisualizationSettings,
+  renderingContext: RenderingContext,
+): CartesianChartDataDensity {
+  const type = "cartesian";
+  const seriesSettingsByDataKey = getDisplaySeriesSettingsByDataKey(
+    seriesModels,
+    stackModels,
+    settings,
+  );
+  const seriesWithSymbols = seriesModels.filter(seriesModel => {
+    const seriesSettings = seriesSettingsByDataKey[seriesModel.dataKey];
+    return ["area", "line"].includes(seriesSettings.display ?? "");
+  });
+  const seriesWithLabels = seriesModels.filter(seriesModel => {
+    const seriesSettings = seriesSettingsByDataKey[seriesModel.dataKey];
+    if (
+      ["area", "bar"].includes(seriesSettings.display ?? "") &&
+      settings["stackable.stack_type"] != null
+    ) {
+      return false;
+    }
+
+    return seriesSettings["show_series_values"];
+  });
+
+  let totalNumberOfDots = 0;
+
+  let totalNumberOfSeriesLabels = 0;
+  let totalNumberOfStackedLabels = 0;
+  let sumOfSeriesLabelWidths = 0;
+  let sumOfStackedSeriesLabelWidths = 0;
+  const fontStyle = {
+    family: renderingContext.fontFamily,
+    weight: CHART_STYLE.seriesLabels.weight,
+    size: CHART_STYLE.seriesLabels.size,
+  };
+
+  dataset.forEach(datum => {
+    totalNumberOfDots += seriesWithSymbols.filter(
+      seriesModel => datum[seriesModel.dataKey] != null,
+    ).length;
+
+    // if we will not be displaying any labels, we do not have to calculate the
+    // label statistics
+    if (
```

Comment:
Seems like we don't need to iterate through the data at all if `!settings["graph.show_values"] ||
      settings["graph.label_value_frequency"] === "all"` so we can move it out of the forEach

Reply:
We still will need to determine how many total dots there will be, and we use that to automatically hide them if there are a lot of them. 

That conditional is only preventing label statistics from being computed since those settings relate to labels.